pkg/monitoring/mimir/service.go

```diff
@@ -0,0 +1,163 @@
+package mimir
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/giantswarm/observability-operator/pkg/common"
+	"github.com/giantswarm/observability-operator/pkg/common/password"
+	"github.com/giantswarm/observability-operator/pkg/common/secret"
+	"github.com/giantswarm/observability-operator/pkg/monitoring/prometheusagent"
+)
+
+const (
+	ingressAuthSecretName = "mimir-gateway-ingress-auth"
+	mimirApiKey           = "mimir-basic-auth"
+	mimirNamespace        = "mimir"
+)
+
+type MimirService struct {
+	client.Client
+	PasswordManager password.Manager
+	common.ManagementCluster
+}
+
+// ConfigureMimir configures the ingress and its authentication (basic auth)
+// to allow prometheus agents to send their data to Mimir
+func (ms *MimirService) ConfigureMimir(ctx context.Context) error {
+	logger := log.FromContext(ctx).WithValues("cluster", ms.ManagementCluster.Name)
+	logger.Info("configuring mimir ingress")
+
+	err := ms.CreateApiKey(ctx, logger)
+	if err != nil {
+		logger.Error(err, "failed to create mimir auth secret")
+		return errors.WithStack(err)
+	}
+
+	err = ms.CreateIngressAuthenticationSecret(ctx, logger)
+	if err != nil {
+		logger.Error(err, "failed to create mimir ingress secret")
+		return errors.WithStack(err)
+	}
+
+	logger.Info("configured mimir ingress")
+
+	return nil
+}
+
+func (ms *MimirService) CreateApiKey(ctx context.Context, logger logr.Logger) error {
+	objectKey := client.ObjectKey{
+		Name:      mimirApiKey,
+		Namespace: mimirNamespace,
+	}
+
+	current := &corev1.Secret{}
+	err := ms.Client.Get(ctx, objectKey, current)
+	if apierrors.IsNotFound(err) {
+		// First all secrets using the password from the mimirApiKey secret are deleted
+		// to ensure that they won't use an outdated password.
+		logger.Info("Deleting old secrets")
+
+		err := secret.DeleteSecret(ingressAuthSecretName, mimirNamespace, ctx, ms.Client)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		clustersList := &clusterv1.ClusterList{}
```

Comment:
Why clustersList and not Cluster? Because it's either clusterList or clusters :D

Reply:
My bad, typo :sweat_smile: